lib/src/v3/connection.dart

```diff
@@ -224,98 +327,30 @@ class PgConnectionImplementation implements PgConnection {
     }
   }
 
-  Future<T> _sendAndWaitForQuery<T extends ServerMessage>(ClientMessage send) {
-    final trace = StackTrace.current;
-
-    return _operationLock.withResource(() {
-      _channel.sink.add(AggregatedClientMessage([send, const SyncMessage()]));
-
-      final completer = Completer<T>();
-      final syncComplete = Completer<void>.sync();
-
-      _pending = _CallbackOperation(this, (message) async {
-        if (message is T) {
-          completer.complete(message);
-        } else if (message is ErrorResponseMessage) {
-          completer.completeError(
-              PostgreSQLException.fromFields(message.fields), trace);
-        } else if (message is ReadyForQueryMessage) {
-          if (!completer.isCompleted) {
-            completer.completeError(
-                StateError('Operation did not complete'), trace);
-          }
-
-          syncComplete.complete();
-        } else {
-          completer.completeError(
-              StateError('Unexpected message $message'), trace);
-        }
-      });
-
-      return syncComplete.future
-          .whenComplete(() => _pending = null)
-          .then((value) => completer.future);
-    });
-  }
-
   @override
-  Future<PgStatement> prepare(Object query, {Duration? timeout}) async {
-    final name = 's/${_statementCounter++}';
-    final description = InternalQueryDescription.wrap(query);
-
-    await _sendAndWaitForQuery<ParseCompleteMessage>(ParseMessage(
-      description.transformedSql,
-      statementName: name,
-      types: description.parameterTypes,
-    ));
-
-    return _PreparedStatement(description, name, this);
+  Future<R> run<R>(Future<R> Function(PgSession session) fn) {
+    return Future.sync(() => fn(this));
```

Comment:
Not sure, so I have to ask: wouldn't this need some kind of exclusion mechanism too? E.g. if the connection is already locked for a transaction in `runTx`, then we shouldn't be sending commands from here.

Reply:
I think it should be fine since accesses to `this` session (like preparing statements or executing them) _will_ be synchronized through the lock. So you can obtain a session via `run` while `runTx` is active, but not actually anything with it until the transaction fonished.

I can add a comment explaining that behavior if that's what we want. Otherwise I can also acquire the lock there to avoid that.

Reply:
I think a description would be fine.